File: Web/Tools/AxMath.js
AxMath.Pi = 3.14159265358979;
AxMath.Pi_d = 3.141592653589793238462;

// a.k.a. Tau
AxMath.Pi2 = AxMath.Pi * 2.0;
AxMath.Pi2_d = AxMath.Pi_d * 2.0;

AxMath.PiHalf = AxMath.Pi / 2.0;
AxMath.PiHalf_d = AxMath.Pi_d / 2.0;

// a.k.a The golden ratio
AxMath.Phi = 1.61803398874989;
AxMath.Phi_d = 1.6180339887498948482;

// Euler's number
AxMath.E = 2.71828182845904;
AxMath.E_d = 2.7182818284590452353;

AxMath.IntMin = Math.trunc(Number.MIN_VALUE);
AxMath.IntMax = Math.trunc(Number.MIN_VALUE);

AxMath.FloatMin = Number.MIN_VALUE;
AxMath.FloatMax = Number.MAX_VALUE;

AxMath.crcPolynomial_Normal = 0x04C11DB7;
AxMath.crcPolynomial_Reversed = 0xEDB88320;
AxMath.crcPolynomial_ReversedReciprocal = 0x82608EDB;

AxMath.crcTable = [];

/**
 * @constructor
 */
function AxMath()
{
}



// Static methods



/**
 * Returns the truncated integer part of the given argument
 * @param {Number} a The argument for which to get truncated value for
 * @return {Number} The truncated integer part of the argument
 */
AxMath.Trunc = function(a)
{
    return Math.trunc(a);
    //return a - a % 1;
    //return parseInt(a);
};

/**
 * Returns the nearest integer value which is not less than the argument
 * @param {Number} a The argument for which to get integer for
 * @return {Number} The nearest integral value which is not less than the argument
 */
AxMath.Ceil = function(a)
{
    return Math.ceil(a);
};

/**
 * Returns the nearest integer value which is not greater than the argument
 * @param {Number} a The argument for which to get integer for
 * @return {Number} The nearest integral value which is not greater than the argument
 */
AxMath.Floor = function(a)
{
    return Math.floor(a);
};

/**
 * Rounds a value to an integer number
 * @param {Number} x The number which to round
 * @return {Number} Returns an integer rounded value of the given number
 */
AxMath.Round = function(x)
{
    return Math.round(x);
};

/**
 * Returns the lesser of two values
 * @param {Number} a A value to compare against
 * @param {Number} b A value to compare against
 * @return {Number} The lesser of the two given values
 */
AxMath.Min = function(a, b)
{
    if (a < b)
        return a;

    return b;
};

/**
 * Returns the greater of two values
 * @param {Number} a A value to compare against
 * @param {Number} b A value to compare against
 * @return {Number} The greater of the two given values
 */
AxMath.Max = function(a, b)
{
    if (a > b)
        return a;

    return b;
};

/**
 * Returns a number limited by a minimum and a maximum boundaries
 * @param {Number} x The number to trim
 * @param {Number} min Minmimum boundary to trim against
 * @param {Number} max Maximum boundary to trim agains
 * @return {Number} The number, trimmed if necessary, by the given boundaries
 */
AxMath.Trim = function(x, min, max)
{
    if (x < min)
        return min;

    if (x > max)
        return max;

    return x;
};

/**
 * Returns a value of the high 16 bits of a given number
 * @param {Number} value The value to get the high 16 bits from
 * @return {Number} The high 16 bits of the given number
 */
AxMath.High = function(value)
{
    return (value >>> 16) & 0xffff;
};

/**
 * Returns a value of the low 16 bits of a given number
 * @param {Number} value The value to get the low 16 bits from
 * @return {Number} The low 16 bits of the given number
 */
AxMath.Low = function(value)
{
    return value & 0xffff;
};

/**
 * Returns a value whose high and low 16 bit parts are constructed by the low 16 bits of two given values
 * @param {Number} high The value from which to form the high 16 bit part of the result
 * @param {Number} low The value from which to form the low 16 bit part of the result
 * @return {Number} A number whose high and low 16 bit parts are constructed by the low 16 bits of two given values
 */
AxMath.HighLow = function(high, low)
{
    return (high << 16) | (low & 0xffff);
};

/**
 * Returns the sign value of a given number
 * @param {Number} x The number whose sign to get
 * @return {Number} Returns 1 if the number is positive, -1 if negative and 0 if it's zero
 */
AxMath.Sign = function(x)
{
    return x > 0 ? 1 : x < 0 ? -1 : 0;
};

/**
 * Returns the absolute value of a given number
 * @param {Number} x The number whose absolute value to get
 * @return {Number} Returns the absolute value of the given number
 */
AxMath.Abs = function(x)
{
    return x >= 0 ? x : -x;
};

/**
 * Raises a number to a given exponent
 * @param {Number} base The base which to raise
 * @param {Number} exponent The power to which the base will be raised
 * @return {Number} Retrns the base raised to the given exponent
 */
AxMath.Power = function(base, exponent)
{
    return Math.pow(base, exponent);
};

/**
 * Calculates the natural exponent, that is the Euler's number, e, to the power of a given exponent
 * @param {Number} exponent The power to which e will be raised
 * @returns {Number} The natural exponent of the given value
 */
AxMath.Exp = function(exponent)
{
    return Math.exp(exponent);
};

/**
 * Calculates the square root of a number
 * @param {Number} x The radicant number whose squre root to calculate 
 * @return {Number} Returns the square root of the given number
 */
AxMath.Sqrt = function(x)
{
    return Math.sqrt(x);
};

/**
 * Calculates the inverse square root of a number, i.e. 1 / Sqrt(x)
 * @param {Number} x The number whose inverse square root to calculate
 * @return {Number} Returns the inverse square root of the given number
 */
AxMath.InvSqrt = function(x)
{
    return 1 / Math.sqrt(x);
};

/**
 * Calculates the logarithm of x to a given base
 * @param {Number} base The base of the logarithm
 * @param {Number} x The argument of the logarithm
 * @returns {Number} The logarithm of x to the given base
 */
AxMath.Log = function(base, x)
{
    return Math.log(x) / Math.log(b);
};

/**
 * Calculates the natural logarithm of x, that is, the logarithm of x to the Euler number, e
 * @param {Number} x The argument of the logarithm
 * @returns {Number} The natural logarithm of x
 */
AxMath.Ln = function(x)
{
    return Math.log(x);
};

/**
 * Calculates the common logarithm of x, that is, the logarithm of x to base 10
 * @param {Number} x The argument of the logarithm
 * @returns {Number} The common logarithm of x
 */
AxMath.Lg = function(x)
{
    return Math.log10(x);
};

/**
 * Calculates the sine of a number
 * @param {Number} x The number in radians whose sine to calculate
 * @return {Number} Returns the sine of the given number
 */
AxMath.Sin = function(x)
{
    return Math.sin(x);
};

/**
 * Calculates the cosine of a number
 * @param {Number} x The number in radians whose cosine to calculate
 * @return {Number} Returns the cosine of the given number
 */
AxMath.Cos = function(x)
{
    return Math.cos(x);
};

/**
 * Calculates the tangent of a number
 * @param {Number} x The number in radians whose tangent to calculate
 * @return {Number} Returns the tangent of the given number
 */
AxMath.Tan = function(x)
{
    return Math.tan(x);
};

/**
 * Calculates the cotangent of a number
 * @param {Number} x The number in radians whose cotangent to calculate
 * @return {Number} Returns the cotangent of the given number
 */
AxMath.Cot = function(x)
{
    return 1.0 / Math.tan(x);
};


/**
 * Calculates the inverse sine of a number
 * @param {Number} x The number in whose arcsine to calculate
 * @return {Number} Returns the arcsine of the given number
 */
AxMath.ArcSin = function(x)
{
    return Math.asin(x);
};

/**
 * Calculates the inverse cosine of a number
 * @param {Number} x The number in whose arccosine to calculate
 * @return {Number} Returns the arccosine of the given number
 */
AxMath.ArcCos = function(x)
{
    return Math.acos(x);
};

/**
 * Calculates the inverse tangent of a number
 * @param {Number} x The number in whose arctangent to calculate
 * @return {Number} Returns the arctangent of the given number
 */
AxMath.ArcTan = function(x)
{
    return Math.atan(x);
};

/**
 * Calculates the inverse cotangent of a number
 * @param {Number} x The number in whose arccotangent to calculate
 * @return {Number} Returns the arccotangent of the given number
 */
AxMath.ArcCot = function(x)
{
    return AxMath.Pi / 2.0 - Math.atan(x);
};

/**
 * Calculates the inverse of the tangent, obtained by two cartesian coordinates
 * Produces a the quadrant-correct angle between the positive x-axis and the vector with the given coordinates
 * @param {Number} x The X coordinate of the vector whose angle to calculate
 * @param {Number} y The Y coordinate of the vector whose angle to calculate
 * @return {Number} Returns the arctangent of the vector with the given coordinates
 */
AxMath.ArcTan2 = function(y, x)
{
    return Math.atan2(y, x);
};



/**
 * Produces a random integer number ranging from zero (including) to a given limit (excluding), i.e. in [0, n)
 * @param {Number} n The maximum limit, between zero and which to generate the random number. The result can not include the limit itself
 * @return {Number} Returns a random number between (and including) zero and the given limit (excluding)
 */
AxMath.RandomInt = function(n)
{
    return Math.trunc(Math.random() * n);
};

/**
 * Produces a random floating point number between zero (including) and 1 (excluding), i.e. in [0, 1)
 * @return {Number} Returns a random number between (and including) zero and 1 (excluding)
 */
AxMath.Random = function()
{
    return Math.random();
};

/**
 * Generates the table used to calculate CRC sums.
 * @param {Number} crcPolynomial The the polynomial used for generating the table
 */
AxMath.CRC32GenerateTable = function(crcPolynomial)
{
    AxMath.crcTable = [];
    
    for (var i = 0; i < 256; i++)
    {
        var remainder = i;
        for(var j = 0; j < 8; j++)
        {
            if (remainder & 1)
            {
                remainder = (remainder >>> 1) ^ crcPolynomial;
            }
            else
            {
                remainder >>>= 1;
            }
        }

        AxMath.crcTable.push(remainder);
    }
};

/**
 * Calculates the CRC sum for a given source, based on a CRC table and a seed
 * @param {Number} source The source for which to calculate a CRC sum
 * @param {Number} size Size of the source in bytes
 * @param {Number} seed The initial seed for calculating the CRC sum
 * @return {Number} Returns a CRC sum for the given source
 */
AxMath.CRC32 = function(source,  size, seed)
{
    var result = ~seed;

    for (var i = 0; i < size; i++)
        result = AxMath.crcTable[((result & 0xFF) ^ source[i])] ^ (result >>> 8);
            
    return ~result;
};



AxMath.CRC32GenerateTable(AxMath.crcPolynomial_Reversed);

File: Engine/Axis/Source/Tools/AxMath.cpp
//---------------------------------------------------------------------------------------------
//	Axis 3D Engine Copyright (C) 2017 Palitri
//
//	This software is released under the MIT License http://palitri.com/axis/license/
//
//	For help and documentation, visit http://palitri.com
//---------------------------------------------------------------------------------------------

#include "AxMath.h"

#include "AxMem.h"

#include <stdlib.h>
#include <time.h>
#include <cmath>
#include <float.h>
#include <limits.h>

const float AxMath::Pi = 3.14159265358979f;
const double AxMath::Pi_d = 3.141592653589793238462;

// a.k.a. Tau
const float AxMath::Pi2 = AxMath::Pi * 2.0f;
const double AxMath::Pi2_d = AxMath::Pi_d * 2.0;

const float AxMath::PiHalf = AxMath::Pi / 2.0f;
const double AxMath::PiHalf_d = AxMath::Pi_d / 2.0;

// a.k.a The golden ratio
const float AxMath::Phi = 1.61803398874989f;
const double AxMath::Phi_d = 1.6180339887498948482;

// Euler's number
const float AxMath::E = 2.71828182845904f;
const double AxMath::E_d = 2.7182818284590452353;

const int AxMath::IntMin = INT_MIN;
const int AxMath::IntMax = INT_MAX;

const float AxMath::FloatMin = FLT_MIN;
const float AxMath::FloatMax = FLT_MAX;

const int AxMath::SHA1SizeBytes = 20;

const unsigned int AxMath::crcPolynomial_Normal = 0x04C11DB7;
const unsigned int AxMath::crcPolynomial_Reversed = 0xEDB88320;
const unsigned int AxMath::crcPolynomial_ReversedReciprocal = 0x82608EDB;

unsigned int AxMath::crcTable[256];

AxMath::Initialization AxMath::initialization;

AxMath::Initialization::Initialization(void)
{
	AxMath::CRC32GenerateTable(AxMath::crcPolynomial_Reversed);
	AxMath::Randomize();
}

AxMath::AxMath(void)
{
}


AxMath::~AxMath(void)
{
}


float AxMath::Trunc(float x)
{
	return (float)((int)x);
}

double AxMath::Trunc(double x)
{
	return (double)((int)x);
}

float AxMath::Floor(float x)
{
	return floorf(x);
}

double AxMath::Floor(double x)
{
	return floor(x);
}

float AxMath::Ceil(float x)
{
	return ceilf(x);
}

double AxMath::Ceil(double x)
{
	return ceil(x);
}

float AxMath::Round(float x)
{
	return floorf(x + 0.5f);
}

double AxMath::Round(double x)
{
	return floor(x + 0.5);
}

int AxMath::Min(int a, int b)
{
	if (a < b)
		return a;

	return b;
}

long long AxMath::Min(long long a, long long b)
{
	if (a < b)
		return a;

	return b;
}

float AxMath::Min(float a, float b)
{
	if (a < b)
		return a;

	return b;
}

int AxMath::Max(int a, int b)
{
	if (a > b)
		return a;

	return b;
}

long long AxMath::Max(long long a, long long b)
{
	if (a > b)
		return a;

	return b;
}

float AxMath::Max(float a, float b)
{
	if (a > b)
		return a;

	return b;
}

int AxMath::Trim(int x, int min, int max)
{
	if (x < min)
		return min;

	if (x > max)
		return max;

	return x;
}

long long AxMath::Trim(long long x, long long min, long long max)
{
	if (x < min)
		return min;

	if (x > max)
		return max;

	return x;
}

float AxMath::Trim(float x, float min, float max)
{
	if (x < min)
		return min;

	if (x > max)
		return max;

	return x;
}

int AxMath::High(int value)
{
	return value >> 16;
}

int AxMath::Low(int value)
{
	return value & 0xffff;
}

int AxMath::HighLow(int high, int low)
{
	return (high << 16) | (low & 0xffff);
}

int AxMath::Sign(int x)
{
	return x > 0 ? 1 : x < 0 ? -1 : 0;
}

float AxMath::Sign(float x)
{
	return x > 0 ? 1.0f : x < 0 ? -1.0f : 0.0f;
}

int AxMath::Abs(int x)
{
	return x >= 0 ? x : -x;
}

float AxMath::Abs(float x)
{
	return x >= 0 ? x : -x;
}

int AxMath::Power(int base, int exponent)
{
	return (int)pow((float)base, exponent);
}

float AxMath::Power(float base, float exponent)
{
	return powf(base, exponent);
}

double AxMath::Power(double base, double exponent)
{
	return pow(base, exponent);
}

float AxMath::Exp(float x)
{
	return exp(x);
}

double AxMath::Exp(double x)
{
	return exp(x);
}

float AxMath::Sqrt(float x)
{
	return sqrtf(x);
}

double AxMath::Sqrt(double x)
{
	return sqrt(x);
}

// TODO: test against 1 / Sqrt (SSE rsqrtss is faster and the compiler might optimize the code)
float AxMath::InvSqrt(float x)
{
    long i;
    float x2, y;
    const float threehalfs = 1.5F;
 
    x2 = x * 0.5f;
    y  = x;
    i  = *(long*)&y;						
    i  = 0x5F3759DF - (i >> 1);				
    y  = *(float*)&i;
    y  = y * (threehalfs - (x2 * y * y));	
 
    return y;
}

float AxMath::Log(float base, float x)
{
	return log(x) / log(base);
}

double AxMath::Log(double base, double x)
{
	return log(x) / log(base);
}


float AxMath::Ln(float x)
{
	return log(x);
}

double AxMath::Ln(double x)
{
	return log(x);
}


float AxMath::Lg(float x)
{
	return log10(x);
}

double AxMath::Lg(double x)
{
	return log10(x);
}


float AxMath::Sin(float x)
{
	return sinf(x);
}

double AxMath::Sin(double x)
{
	return sin(x);
}


float AxMath::Cos(float x)
{
	return cosf(x);
}

double AxMath::Cos(double x)
{
	return cos(x);
}


float AxMath::Tan(float x)
{
	return tanf(x);
}

double AxMath::Tan(double x)
{
	return tan(x);
}


float AxMath::Cot(float x)
{
	return 1.0f / tanf(x);
}

double AxMath::Cot(double x)
{
	return 1.0 / tan(x);
}


float AxMath::ArcSin(float x)
{
	return asinf(x);
}

double AxMath::ArcSin(double x)
{
	return asin(x);
}


float AxMath::ArcCos(float x)
{
	return acosf(x);
}

double AxMath::ArcCos(double x)
{
	return acos(x);
}


float AxMath::ArcTan(float x)
{
	return atanf(x);
}

double AxMath::ArcTan(double x)
{
	return atan(x);
}


float AxMath::ArcCot(float x)
{
	return AxMath::Pi / 2.0f - atanf(x);
}

double AxMath::ArcCot(double x)
{
	return AxMath::PiHalf_d - atan(x);
}


float AxMath::ArcTan2(float y, float x)
{
	return atan2f(y, x);
}

double AxMath::ArcTan2(double y, double x)
{
	return atan2(y, x);
}

unsigned char AxMath::ROL(unsigned char value, int bits)
{
	return (value << bits) | (value >> (8 - bits));
}

unsigned short AxMath::ROL(unsigned short value, int bits)
{
	return (value << bits) | (value >> (16 - bits));
}

unsigned int AxMath::ROL(unsigned int value, int bits)
{
	return (value << bits) | (value >> (32 - bits));
}

unsigned long long AxMath::ROL(unsigned long long value, int bits)
{
	return (value << bits) | (value >> (64 - bits));
}

unsigned char AxMath::ROR(unsigned char value, int bits)
{
	return (value >> bits) | (value << (8 - bits));
}

unsigned short AxMath::ROR(unsigned short value, int bits)
{
	return (value >> bits) | (value << (16 - bits));
}

unsigned int AxMath::ROR(unsigned int value, int bits)
{
	return (value >> bits) | (value << (32 - bits));
}

unsigned long long AxMath::ROR(unsigned long long value, int bits)
{
	return (value >> bits) | (value << (64 - bits));
}

int AxMath::Random(int n)
{
	return rand() % n;
}

float AxMath::Random()
{
	return (float)rand() / (float)RAND_MAX;
}

void AxMath::Randomize()
{
	srand((unsigned int)time(NULL));
}

void AxMath::CRC32GenerateTable(unsigned int crcPolynomial)
{
	for (int i = 0; i < 256; i++)
	{
		unsigned int remainder = i;
		for(int j = 0; j < 8; j++)
		{
			if (remainder & 1)
			{
				remainder = (remainder >> 1) ^ crcPolynomial;
			}
			else
			{
				remainder >>= 1;
			}
		}

		AxMath::crcTable[i] = remainder;
	}
}

unsigned int AxMath::CRC32(const void *source, unsigned int size, unsigned int seed)
{
    unsigned int result = ~seed;

    for (unsigned int i = 0; i < size; i++)
        result = AxMath::crcTable[((result & 0xFF) ^ ((unsigned char*)source)[i])] ^ (result >> 8);
            
    return ~result;
}

// CRC16-CCITT, polynomial 0x1021 (0b1000000100001), that is x^16 + x^12 + x^5 + 1
unsigned short AxMath::CRC16(const void *source, unsigned int size, unsigned short seed)
{
	unsigned short result = seed;

	for (unsigned int i = 0; i < size; i++)
	{
		unsigned char x = result >> 8 ^ ((unsigned char*)source)[i];
		x ^= x >> 4;
		result = (result << 8) ^ ((unsigned short)(x << 12)) ^ ((unsigned short)(x << 5)) ^ ((unsigned short)x);
	}

	return result;
}

unsigned short AxMath::IterativeCRC16Init(unsigned short seed)
{
	return seed;
}

unsigned short AxMath::IterativeCRC16Iterate(unsigned short crc16, unsigned char data)
{
	unsigned char x = crc16 >> 8 ^ data;
	x ^= x >> 4;
	crc16 = (crc16 << 8) ^ ((unsigned short)(x << 12)) ^ ((unsigned short)(x << 5)) ^ ((unsigned short)x);
	
	return crc16;
}


void AxMath::SHA1(void *result, const void *data, int size)
{
	unsigned int h0 = 0x67452301;
	unsigned int h1 = 0xEFCDAB89;
	unsigned int h2 = 0x98BADCFE;
	unsigned int h3 = 0x10325476;
	unsigned int h4 = 0xC3D2E1F0;


	unsigned long long sizeBits = size * 8;
	unsigned long long bufferSizeBits = sizeBits;

	// Calculate size of buffer in bits. That is original data size plus one bit with value "1" added at the end of data, plus zero bits so that total buffer size is divisible by 512, plus 64 bits containing data length
	bufferSizeBits += 1 + 64;
	unsigned long long mod = bufferSizeBits % 512;
	if (mod != 0)
		bufferSizeBits += 512 - mod;

	unsigned int bufferSizeBytes = (unsigned int)(bufferSizeBits / 8);

	// Create a buffer, copy original data, append bit "1", append bits "0", so that buffer size in bits is devisible by 448, append 64 bits containing data length, so total buffer size in bits is divisible by 512
	char *buffer = new char[bufferSizeBytes];
	AxMem::Copy(buffer, data, size);
	buffer[size] = 0x80;
	AxMem::Fill(&(buffer[size + 1]), bufferSizeBytes - size - 1, "\0", 1);
	AxMem::CopyReversedByteOrder(&buffer[bufferSizeBytes - 8], &sizeBits, 8);


	unsigned int blocksCount = bufferSizeBits / 512;
	unsigned int blockWords[80];

	for (unsigned int blockIndex = 0; blockIndex < blocksCount; blockIndex++)
	{
		for (int i = 0; i < 16; i++)
			AxMem::CopyReversedByteOrder(&blockWords[i], &buffer[blockIndex * 64 + i * 4], 4);

		for (int i = 16; i < 80; i++)
			blockWords[i] = AxMath::ROL(blockWords[i - 3] ^ blockWords[i - 8] ^ blockWords[i - 14] ^ blockWords[i - 16], 1);

		unsigned int a = h0;
		unsigned int b = h1;
		unsigned int c = h2;
		unsigned int d = h3;
		unsigned int e = h4;

		for (int i = 0; i < 80; i++)
		{
			unsigned int f, k;
			if (i >= 60)
			{
				f = b ^ c ^ d;
				k = 0xCA62C1D6;
			}
			else if (i >= 40)
			{
				f = (b & c) | (b & d) | (c & d);
				k = 0x8F1BBCDC;
			}
			else if (i >= 20)
			{
				f = b ^ c ^ d;
				k = 0x6ED9EBA1;
			}
			else
			{
				f = (b & c) | ((~b) & d);
				k = 0x5A827999;
			}

			unsigned int temp = AxMath::ROL(a, 5) + f + e + k + blockWords[i];
			e = d;
			d = c;
			c = AxMath::ROL(b, 30);
			b = a;
			a = temp;
		}

		h0 += a;
		h1 += b;
		h2 += c;
		h3 += d;
		h4 += e;
	}

	delete[] buffer;

	AxMem::CopyReversedByteOrder((char*)result, &h0, 4);
	AxMem::CopyReversedByteOrder((char*)result + 4, &h1, 4);
	AxMem::CopyReversedByteOrder((char*)result + 8, &h2, 4);
	AxMem::CopyReversedByteOrder((char*)result + 12, &h3, 4);
	AxMem::CopyReversedByteOrder((char*)result + 16, &h4, 4);
}